### PR TITLE
Ecal CC timing reconstruction deployment

### PR DESCRIPTION
**Pull request by jking79**
#### PR description:

PR for deployment of ECAL CC timing reconstruction part 2 

encoding the non corrected CC in the 8 bit Timing Error:
More details https://indico.cern.ch/event/1210633/contributions/5091427/attachments/2527245/4347248/ECAL_2022_timeError.pdf
Added encoding constants declaration to EcalConstants.h
Added functions to set and return non corrected time to EcalUncalibratedRecHit.h & [EcalUncalibratedRecHit.cc](http://ecaluncalibratedrechit.cc/)
Added function to return non corrected time to EcalRecHit.h

Modified ecalMultiFitUncalibRecHit_cfi.py to make crossCorrelationMethod default timing reconstruction
More details: https://indico.cern.ch/event/1114550/contributions/4887537/attachments/2451361/4200675/ECAL_DPG_25May22.pdf

#### PR validation:

scram b runtests and runTheMatrix.py -l limited -i all --ibeos test where successfully completed.

### CODE REVIEW

**Comment by makortel** [comment]
Please change the commit history such that the `.gitignore` file is not touched at all (e.g. squash 9f51e59b06dc5aabfe00ce48be2e5d6c26ae821d, a92d62ad1dd29e533d8490f9d2c1ef0007d7a704, and ddcdc962bd10f9cab801f7958d7db0e4710f5470 into fd955782c06ded7a23bc694139195603215c224b).

**Comment by civanch** [comment]
please test

File: DataFormats/EcalDigi/interface/EcalConstants.h [hunk 1]
```diff
@@ -4,6 +4,15 @@
 
 #include "FWCore/Utilities/interface/HostDeviceConstant.h"
 
+// constants used to encode difference between corrected cc time and noncorrected cc time
+// into a uInt8 value to be stored in the 8 bits set aside for the jitter error ( previously unused )
+namespace ecalcctiming {
+  static constexpr const float clockToNS = 25.0;          // Convert clock units to ns
+  static constexpr const float nonCorrectedSlope = 1.2;   // estimates nonCorreted time from corrected time
+  static constexpr const float encodingOffest = 0.32;     // offsets difference in time using clock units
+  static constexpr const float encodingValue = 398.4375;  // encodes time difference into 0 - 255 int range
```
Comment (clacaputo) [inline DataFormats/EcalDigi/interface/EcalConstants.h:13]: For my education, why `encodingValue` that encodes the time difference `t_w_ootc - t_wo_ootc` is fixed and not Era/Condition dependent?

File: RecoLocalCalo/EcalRecProducers/test/testEcalUncalibRechitProducerWithCC_cfg.py [hunk 2]
```diff
@@ -47,15 +47,15 @@
...
-process.ecalMultiFitUncalibRecHit.algoPSet = cms.PSet( 
-      # for crossCorrelationMethod
-      timealgo = cms.string( "crossCorrelationMethod" ),
-    #   timealgo = cms.string( "RatioMethod" ),  
-    #   crossCorrelationStartTime = cms.double(-15),
-    #   crossCorrelationStopTime = cms.double(25),
-    #   crossCorrelationTargetTimePrecision = cms.double(0.01),
-    #   crossCorrelationTargetTimePrecisionForDelayedPulses = cms.double(0.05),
-)     
+#process.ecalMultiFitUncalibRecHit.algoPSet = cms.PSet(
```
Comment (clacaputo) [inline RecoLocalCalo/EcalRecProducers/test/testEcalUncalibRechitProducerWithCC_cfg.py:50]: Can't you simple remove all the block commented?

File: DataFormats/EcalDigi/interface/EcalConstants.h (hunk 1 above)
Comment (makortel) [reply DataFormats/EcalDigi/interface/EcalConstants.h:13]: By the way, constants defined in a namespace (in contrast to a class) should be defined as `inline constexpr`. `static constexpr` tells the compiler to create a separate copy for each file that includes the header.

**Comment by clacaputo** [comment]
Reco differences in `EcalRecHitsSorted_reducedEcalRecHitsEB__RECO_obj_obj_time` `EcalRecHitsSorted_ecalRecHit_EcalRecHitsEB_RECO_obj_obj_time` should be a desired effect of this PR. Could you please confirm @jking79 ?

c_EcalRecHitsSorted_reducedEcalRecHitsEB__RECO_obj_obj_time

c_EcalRecHitsSorted_ecalRecHit_EcalRecHitsEB_RECO_obj_obj_time

**Comment by jking79** [comment]
That is correct.  The reco differences in ECAL RecHit times are the desired effect of this PR.

**Comment by jking79** [comment]
The encodingValue is only dependent on the range of differences that can be encoding at a particular precision with an 8 bit value. This difference is the difference in the CC time reconstruction with and with out an out of time amplitude correction that mitigates the impact of out of time pile up. This correction, which is dependent on the amplitudes derived from the ECAL multifit algorithm, has been show to have a bias at large delay times such as those that can be observed in events with long lived particles. This encoding stores a correction that will be applied to the calibrated rechit time to compensate for that bias.  

As for the block comment in testEcalUncalibRechitProducerWithCC_cfg.py, that is an oversight on my part.  Should I remove them and push the changes?

**Comment by clacaputo** [comment]
> As for the block comment in testEcalUncalibRechitProducerWithCC_cfg.py, that is an oversight on my part. Should I remove them and push the changes?

Hi @jking79 , please remove the commented code in `testEcalUncalibRechitProducerWithCC_cfg.py`. Thanks.

> This encoding stores a correction that will be applied to the calibrated rechit time to compensate for that bias.

It means that this bias will always be the same, despite all the possible conditions' changes in the ECAL. Do I understand it correctly?

**Comment by jking79** [comment]
Hello,

That is correct, there is no dependance on condition changes in the ECAL for the constants used to encoded the correction for the bias.

Excelsior,

Jack

**Comment by civanch** [comment]
+1

**Comment by clacaputo** [comment]
+reconstruction

**Comment by perrotta** [comment]
please test
(just to have a look at the outpput of the static checks, here missing)

**Comment by perrotta** [comment]
+1

**Comment by perrotta** [comment]
merge

**Pull request merged**